streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -157,25 +158,44 @@ private void startRestoration(final Map<TopicPartition, StateRestorer> initializ
             if (restorer.checkpoint() != StateRestorer.NO_CHECKPOINT) {
-                consumer.seek(restorer.partition(), restorer.checkpoint());
-                logRestoreOffsets(restorer.partition(),
-                        restorer.checkpoint(),
-                        endOffsets.get(restorer.partition()));
-                restorer.setStartingOffset(consumer.position(restorer.partition()));
+                consumer.seek(restoringPartition, restorer.checkpoint());
+                logRestoreOffsets(restoringPartition,
+                                  restorer.checkpoint(),
+                                  endOffsets.get(restoringPartition));
+                restorer.setStartingOffset(consumer.position(restoringPartition));
                 restorer.restoreStarted();
             } else {
-                consumer.seekToBeginning(Collections.singletonList(restorer.partition()));
+                consumer.seekToBeginning(Collections.singletonList(restoringPartition));
                 needsPositionUpdate.add(restorer);
             }
         }
 
         for (final StateRestorer restorer : needsPositionUpdate) {
-            final long position = consumer.position(restorer.partition());
-            logRestoreOffsets(restorer.partition(),
-                              position,
-                              endOffsets.get(restorer.partition()));
-            restorer.setStartingOffset(position);
+            final TopicPartition restoringPartition = restorer.partition();
+            final StreamTask task = active.restoringTaskFor(restoringPartition);
+
+            // If checkpoint does not exist it means the task was not shutdown gracefully before;
+            // and in this case if EOS is turned on we should wipe out the state and re-initialize the task
+            if (task.eosEnabled) {
+                log.info("No checkpoint found for task {} state store {} changelog {} with EOS turned on. " +
+                    "Reinitializing the task and restore its state from the beginning.", task.id, restorer.storeName(), restoringPartition);
+                needsInitializing.remove(restoringPartition);
+
+                initialized.put(restoringPartition, restorer);
+                restorer.setCheckpointOffset(consumer.position(restoringPartition));
+                task.reinitializeStateStoresForPartitions(restoringPartition);
+            } else {
+                log.info("Restoring task {}'s state store {} from beginning of the changelog {} ", task.id, restorer.storeName(), restoringPartition);
+
+                final long position = consumer.position(restoringPartition);
+                logRestoreOffsets(restoringPartition,
+                                  position,
+                                  endOffsets.get(restoringPartition));
+                restorer.setStartingOffset(position);
+            }
+
             restorer.restoreStarted();
```

Comment:
This is a difference to the original PR. I think it's a bug in `trunk` and `2.0`. Cf. https://github.com/apache/kafka/pull/5430/files#diff-46ed6d177221c8778965ecb1b6657be3R232

Reply:
Only if EOS is not turned on, we will still continue restoring but from the starting position, and hence calling `restorer.restoreStarted();`. In the other case, we call put the restorer back to `initialized` `task.reinitializeStateStoresForPartitions` which will then call `register` again to add the partition back to `needsInitializing.add(restorer.partition());` which will then be re-initialized in the next iteration. I think the logic is still correct.

Reply:
If I understand you comment correctly, you say that the code in `trunk`/`2.0` is correct while here it's incorrect and we should call `restorer.restoreStarted();` only for the non-EOS case here (and that the EOS case is covered above in line https://github.com/apache/kafka/pull/5657/files#diff-46ed6d177221c8778965ecb1b6657be3R168)

Reply:
I double check this, too. If we don't call `restorer.restoreStarted();` we don't call it all. However, if we call it here, it's incorrect, too, because within `task.reinitializeStateStoresForPartitions()` we create a restorer object.

Frankly, I am not 100% sure what the difference between `trunk` and `1.0` is -- we have changed some logic such that in this branch, after re-initializing, we never enter this method again, and thus, the call to `restorer.restoreStarted();` in line 166/168 is never triggered (ie, we need to make the call here). In `trunk` however, `startRestoration` is called once again and thus we don't need to call `restorer.restoreStarted();` after task.reinitializeStateStoresForPartitions()`.

Thoughts?